### PR TITLE
[HAL] Implement IPI support functions

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Implement IPI support functions in hal.

### CODE REVIEW

File: hal/halx86/generic/up.c [hunk 1]
```diff
@@ -32,3 +32,22 @@ HalStartNextProcessor(
...
+{
+}
+
+VOID
+NTAPI
+HalSendSoftwareInterrupt(
+    _In_ KAFFINITY TargetSet,
+    _In_ KIRQL Irql)
+{
+}
```
Comment (binarymaster) [inline hal/halx86/generic/up.c:51]:
```suggestion
VOID
NTAPI
HalSendNMI(
    _In_ KAFFINITY TargetSet)
{
    NOTHING;
}

VOID
NTAPI
HalSendSoftwareInterrupt(
    _In_ KAFFINITY TargetSet,
    _In_ KIRQL Irql)
{
    NOTHING;
}
```

**Review by DarkFire01 — APPROVED** [review]
Had to test with what I did for x86 so it took a bit! 

Worked great 👍

**Review by HBelusca — APPROVED** [review]

File: hal/halx86/apic/apicsmp.c [hunk 2]
```diff
@@ -138,3 +130,172 @@ ApicStartApplicationProcessor(
...
+    {
+        Vector = APC_VECTOR;
+    }
+    else if (Irql == DISPATCH_LEVEL)
+    {
+        Vector = DISPATCH_VECTOR;
+    }
+    else
+    {
+        ASSERT(FALSE);
```
Comment (HBelusca) [inline hal/halx86/apic/apicsmp.c:257]: It's not possible to have other Irqls given to that function?

Comment (tkreuzer) [reply hal/halx86/apic/apicsmp.c:257]: No, that are the only 2 software interrupts.

File: hal/halx86/generic/up.c (hunk 1 above)
Comment (tkreuzer) [reply hal/halx86/generic/up.c:51]: Done

File: hal/halx86/apic/apicp.h [hunk 3]
```diff
@@ -151,7 +151,7 @@ typedef enum _APIC_DSH
     APIC_DSH_Destination,
     APIC_DSH_Self,
     APIC_DSH_AllIncludingSelf,
-    APIC_DSH_AllExclusingSelf
+    APIC_DSH_AllExcludingSelf
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply hal/halx86/apic/apicp.h:154]: Done

**Pull request merged**